pkg/ingester/ingester.go

```diff
@@ -385,6 +393,15 @@ func (i *Ingester) query(ctx context.Context, from, through model.Time, matchers
 			Values: values,
 		})
 		queriedSamples += len(values)
+		queriedSeries++
+
+		if queriedSamples > i.cfg.MaxSamplesPerQuery {
+			return errors.New("exceeded maximum number of samples in a query")
+		}
+		if queriedSeries > i.cfg.MaxSeriesPerQuery {
+			return errors.New("exceeded maximum number of series in a query")
+		}
```

Comment:
Can you put this in to state.forSeriesMatching so we don't use all the memory before checking...